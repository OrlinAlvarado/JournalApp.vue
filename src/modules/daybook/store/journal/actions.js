import journalApi from '@/api/journalApi';

// export const myAction = async ({ commit }) => {
    
// }

export const loadEntries = async ({ commit }) => {
    const { data } = await journalApi.get('/entries.json')
    
    if( !data ){
        commit('setEntries', [])
        return
    }
    
    const entries = []
    
    for(let id of Object.keys( data )){
        entries.push({
            id,
            ...data[id]
        })
    }
    
    commit('setEntries', entries)
}

export const updateEntry = async ({ commit }, entry) => {
    const { date, text, picture } =  entry
    
    const dataToSave = {
        date,
        picture,
        text
    }
    
    await journalApi.put(`/entries/${ entry.id }.json`, dataToSave)
    
    dataToSave.id = entry.id 
    commit('updateEntry', { ...dataToSave })
}

export const createEntry = async ({ commit }, entry) => {
    const { date, text, picture } =  entry
    
    const dataToSave = {
        date,
        picture,
        text
    }
    
    const { data } = await journalApi.post(`/entries.json`, dataToSave)
        
    commit('addEntry', { id: data.name, ...dataToSave })
    
    return data.name
}

export const deleteEntry = async ({ commit }, id) => {
    await journalApi.delete(`/entries/${ id }.json`)
    
    commit('deleteEntry', id)
}