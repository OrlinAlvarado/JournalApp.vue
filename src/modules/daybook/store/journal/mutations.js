

// export const myMutation = ({ commit }) => {
    
// }

export const setEntries = ( /* state */) => {
    
}

export const updateEntry = ( /* state */) => {
    
}

export const addEntry = ( /* state */) => {
    
}