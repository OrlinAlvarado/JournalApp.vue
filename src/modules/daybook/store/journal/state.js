

export default () => ({
    isLoading: true,
    entries: [
        {
            id: '1',
            date: new Date().toDateString(),
            text: 'Occaecat magna eu sunt ipsum eiusmod elit ex dolor mollit ipsum laboris. Nostrud reprehenderit consequat proident excepteur reprehenderit exercitation. Excepteur est adipisicing dolore Lorem id cupidatat anim officia et voluptate tempor labore eiusmod dolor. Lorem veniam occaecat occaecat deserunt tempor sit fugiat.',
            picture: null
        },
        {
            id: '2',
            date: new Date().toDateString(),
            text: 'Ea ullamco Lorem consequat quis reprehenderit velit labore nisi mollit do velit. Excepteur cupidatat pariatur voluptate culpa nostrud deserunt ea eu consequat id est. Ipsum irure elit aute esse.',
            picture: null
        },
        {
            id: '3',
            date: new Date().toDateString(),
            text: 'Ad anim cupidatat ad ad mollit tempor mollit do cillum aliquip quis est eiusmod deserunt. Consectetur veniam elit est laboris incididunt voluptate voluptate ut fugiat do quis tempor in magna. Duis commodo cillum officia id adipisicing enim do et ad laborum dolore excepteur. Enim officia quis in aliquip labore labore mollit proident. Eiusmod qui sunt velit irure Lorem. Laborum minim laborum aliqua aliqua minim dolore irure laboris.',
            picture: null
        },
    ]
})