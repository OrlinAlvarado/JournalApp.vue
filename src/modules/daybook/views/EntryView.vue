<template>
    <template v-if="entry">
        <div class="entry-title d-flex justify-content-between p-2">
            <div>
                <span class="text-success fs-3 fw-bold mx-2">{{ day }}</span>
                <span class="fs-3 fw-bold">{{ month }}</span>
                <span class="mx-2 fs-3 fw-bold">{{ yearDay }}</span>
            </div>
            
            <div>
                <button class="btn btn-danger mx-2">
                    Borrar
                    <i class="fa fa-trash-alt"></i>
                </button>
                <button class="btn btn-primary">
                    Borrar
                    <i class="fa fa-upload"></i>
                </button>
            </div>
        </div>
        
        <hr>
        
        <div class="d-flex flex-column px-3 h-75">
            <textarea
                rows="4"
                placeholder="¿Que sucedió hoy?"
                v-model="entry.text"
            >
                
            </textarea>
        </div>
        
        <img 
            src="https://www.lifeder.com/wp-content/uploads/2020/11/paisaje-natural-y-antropico.jpg" 
            alt="entry-picture"
            class="img-thumbnail"
        >
    </template>
    
    <Fab 
        icon="fa-save"
    />
    
</template>

<script>
import { defineAsyncComponent } from 'vue'
import { mapGetters } from 'vuex'

import getDayMothYear from '../helpers/getDayMothYear'
    
export default {
    props: {
        id: {
            type: String,
            required: true
        }
    },
    components: {
        Fab: defineAsyncComponent(() => import('../components/Fab.vue'))
    },
    data(){
        return {
            entry: null
        }
    },
    computed: {
        ...mapGetters('journal', ['getEntryById']),
        day(){
            const { day } = getDayMothYear( this.entry.date )
            return day
        },
        month(){
            const { month } = getDayMothYear( this.entry.date )
            return month
        },
        yearDay(){
            const { yearDay } = getDayMothYear( this.entry.date )
            return yearDay
        },
    },
    methods: {
        loadEntry(){
            const entry  = this.getEntryById( this.id )
            if( !entry ) return this.$router.push({ name: 'no-entry'})
            
            this.entry = entry
        }
    },
    created(){
       this.loadEntry()
    },
    watch: {
        id(){
            this.loadEntry()
        }
    }
    
}
</script>

<style lang="scss" scoped>

textarea{
    font-size: 20px;
    border: none;
    
    &:focus{
        outline: none;
    }
}

img {
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;  
  box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
}
</style>