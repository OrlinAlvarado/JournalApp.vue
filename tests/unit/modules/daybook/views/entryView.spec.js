import { createStore } from 'vuex'
import { shallowMount } from '@vue/test-utils'

import Swal from 'sweetalert2'

import { journalState } from '../../../mock-data/test-journal-state'
import journal from '@/modules/daybook/store/journal'

import EntryView from '@/modules/daybook/views/EntryView.vue'

const createVuexStore = ( initialState ) =>  
    createStore({
        modules: {
            journal: {
                ...journal,
                state: { ...initialState }
            }
        }
    }
)

jest.mock('sweetalert2', () => ({
   fire: jest.fn(),
   showLoading: jest.fn(), 
   close: jest.fn()
}))

describe('Pruebas en el EntryView', () => {
    const store = createVuexStore( journalState )
    store.dispatch = jest.fn()
    const mockRouter = {
        push: jest.fn()
    }
    
    let wrapper 
    
    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount( EntryView , {
            props: {
                id: '-MkggXlcRyiHVJxQ5A6z'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })
        
    })
    
    test('debe de sacar al usuario porque el id no existe', () => {
        const  wrapper = shallowMount( EntryView , {
            props: {
                id: 'Este id no existe'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        });
        
        expect( mockRouter.push ).toHaveBeenCalledWith({ name: 'no-entry'})
        
        
    })
    
    test('debe de mostrar la entrada correctamente', () => {
        expect(wrapper.html()).toMatchSnapshot()
        expect( mockRouter.push ).not.toHaveBeenCalled()
    })
    
    test('debe de borrar la entrada y salir', async (done) => {
        Swal.fire.mockReturnValueOnce( Promise.resolve({ isConfirmed: true }))
        
        await wrapper.find('.btn-danger').trigger('click')
        
        expect( Swal.fire ).toHaveBeenCalledWith({
            title: '¿Está seguro?',
            text: 'Una vez borrado, no se puede recuperar',
            showDenyButton: true,
            confirmButtonText: 'Si, estoy seguro'
        })
        
        setTimeout(() => {
            expect( store.dispatch ).toHaveBeenCalledWith('journal/deleteEntry', '-MkggXlcRyiHVJxQ5A6z')
            expect(mockRouter.push).toHaveBeenCalled()
            done()
        }, 1);
    })
    
    
    
})
