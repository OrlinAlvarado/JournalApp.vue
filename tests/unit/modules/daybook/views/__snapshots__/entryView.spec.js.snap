// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold mx-2">22</span><span class="fs-3 fw-bold">Septiembre</span><span class="mx-2 fs-3 fw-bold">2021, Miércoles</span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea rows="4" placeholder="¿Que sucedió hoy?">

            </textarea></div><img src="https://res.cloudinary.com/orlinalvarado/image/upload/v1632322156/fnhs5y4gc3cqbvuybusg.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
